cmd/check_spec_utils.go

```diff
@@ -16,18 +16,314 @@ package cmd
+	errBlockNotIdempotent          = errors.New("Multiple calls with the same hash don't return the same block")
+	errBlockTip                    = errors.New("Unspecified block_identifier doesn't give the tip block")
+	errRosettaConfigNoConstruction = errors.New("No construction element in Rosetta config")
+)
+
+type checkSpecAPI string
+type checkSpecRequirement string
+type checkSpecStatus string
+
+const (
+	networkList    checkSpecAPI = "/network/list"
+	networkOptions checkSpecAPI = "/network/options"
+	networkStatus  checkSpecAPI = "/network/status"
+	accountBalance checkSpecAPI = "/account/balance"
+	accountCoins   checkSpecAPI = "/account/coins"
+	block          checkSpecAPI = "/block"
+	errorObject    checkSpecAPI = "error object"
+	modes          checkSpecAPI = "modes"
+
+	networkIDs      checkSpecRequirement = "network_identifiers is required"
+	offlineMode     checkSpecRequirement = "endpoint should work in offline mode"
+	staticNetworkID checkSpecRequirement = "network_identifier must be static"
+	version         checkSpecRequirement = "field version is required"
+	allow           checkSpecRequirement = "field allow is required"
+
+	currentBlockID   checkSpecRequirement = "current_block_identifier is required"
+	currentBlockTime checkSpecRequirement = "current_block_timestamp is required"
+	genesisBlockID   checkSpecRequirement = "genesis_block_identifier is required"
+	peers            checkSpecRequirement = "field peers is required"
+
+	blockID    checkSpecRequirement = "block_identifier is required"
+	balances   checkSpecRequirement = "field balances is required"
+	coins      checkSpecRequirement = "field coins is required"
+	idempotent checkSpecRequirement = "same hash should return the same block"
+	defaultTip checkSpecRequirement = "tip should be returned if block_identifier is not specified"
+
+	errorCode    checkSpecRequirement = "error code is required"
+	errorMessage checkSpecRequirement = "error message is required"
+	diffURLs     checkSpecRequirement = "offline_url should be different from offline_url and not empty"
+
+	checkSpecSuccess checkSpecStatus = "Success"
+	checkSpecFailure checkSpecStatus = "Failure"
 )
 
-func verifyBlockIdentifier(blockID *types.BlockIdentifier) error {
-	if blockID != nil && blockID.Hash != "" && blockID.Index >= 0 {
+type checkSpecOutput struct {
+	api        checkSpecAPI
+	validation map[checkSpecRequirement]checkSpecStatus
+}
+
+func validateBlockIdentifier(blockID *types.BlockIdentifier) error {
+	if blockID == nil {
+		return errBlockIdentifierNullPointer
+	}
+
+	if blockID.Hash == "" {
+		return errBlockIdentifierEmptyHash
+	}
+
+	if blockID.Index < 0 {
+		return errBlockIdentifierNegativeIndex
+	}
+
+	return nil
+}
+
+func validateTimestamp(time int64) error {
+	if time >= 0 {
 		return nil
 	}
 
-	return errBlockIdentifier
+	return errTimestampNegative
+}
+
+func validatePeers(peers []*types.Peer) error {
+	for _, p := range peers {
+		if p == nil {
```

Comment:
General good coding practice for these kind of checks - if you observer you see that you are checking at a lot of places in this file that something is empty and/or null. So a good way is to have a helper function which does that. For example create these functions
```
func IsNullOrEmpty(someString string) bool {
    if isNull(someString) || isEmpty(someString) {
        return true;
    }
}

func isNull(someString string) bool {...}
func isEmpty(someString string) bool {...}

```

These function could be in some kind of helper/util file. This will make sure in case these logic change you don't need to replace anywhere and you can re-use lot of code.